netperf_tool/src/client_node.cpp

```diff
@@ -0,0 +1,146 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "client_node.hpp"
+
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "netperf_tool_interfaces/msg/bytes.hpp"
+#include "netperf_tool_interfaces/srv/get_results.hpp"
+
+#include "rclcpp/executors.hpp"
+#include "rclcpp/node.hpp"
+#include "rclcpp/serialization.hpp"
+#include "rclcpp/serialized_message.hpp"
+
+#include "utils.hpp"
+
+namespace netperf_tool
+{
+
+using namespace std::chrono_literals;
+
+ClientNode::ClientNode(
+  const rclcpp::NodeOptions & options,
+  const rclcpp::QoS & pub_qos,
+  size_t array_size,
+  std::chrono::nanoseconds target_pub_period,
+  std::chrono::nanoseconds server_timeout)
+: Node("client", "netperf_tool", options),
+  target_pub_period_{target_pub_period},
+  server_timeout_{server_timeout}
+{
+  pub_ = this->create_publisher<netperf_tool_interfaces::msg::Bytes>("test_topic", pub_qos);
+  // always keep a message with the data preallocated
+  msg_to_publish_.data.resize(array_size);
+  msg_to_publish_.id = 0;
+  rclcpp::SerializedMessage serialized_msg;
+  rclcpp::Serialization<netperf_tool_interfaces::msg::Bytes> serializer;
+  serializer.serialize_message(&msg_to_publish_, &serialized_msg);
+  serialized_msg_size_ = serialized_msg.size();
+  client_ = this->create_client<netperf_tool_interfaces::srv::GetResults>("get_results");
+}
+
+bool
+ClientNode::wait_for_server()
+{
+  return client_->wait_for_service(server_timeout_);
+}
+
+void
+ClientNode::start_publishing()
+{
+  timer_ = this->create_wall_timer(
+    target_pub_period_,
+    std::bind(&ClientNode::pub_next_msg, this));
+}
+
+void
+ClientNode::stop_publishing()
+{
+  if (timer_) {
+    timer_->cancel();
+  }
+}
+
+void
+ClientNode::sync_with_server(rclcpp::Executor & exec)
+{
+  if (!pub_->wait_for_all_acked(server_timeout_)) {
+    RCLCPP_WARN(this->get_logger(), "Some messages were not acknowledged by the netperf server ...");
+  }
+  if (!client_->service_is_ready()) {
+    RCLCPP_ERROR(
+      this->get_logger(), "Perf server is not available anymore, cannot get the results back");
+    return;
+  }
+  auto req = std::make_shared<netperf_tool_interfaces::srv::GetResults::Request>();
+  req->publisher_gid = this->get_stringified_pub_gid();
+  req->messages_total = this->collected_info_.message_ids.size();
+  auto future = client_->async_send_request(req);
+  exec.spin_until_future_complete(future);
+  statistics_ = future.get();
+}
+
+ClientResults
+ClientNode::extract_results()
+{
+  ClientResults results;
+  if (!statistics_) {
+    throw std::runtime_error{"you must first completely run the client before getting results"};
+  }
+  results.statistics = *statistics_;
+  statistics_.reset();
+  {
+    std::lock_guard guard{collected_info_mutex_};
+    results.collected_info = std::move(collected_info_);
```

Comment:
I'm wondering if the mutex is necessary. In what scenario can we run into a race? IIUC, the timer should already be canceled so we shouldn't be accessing this in another thread by this point. In the event that the user prematurely terminates the program, I think `statistics_` would not be set anyways and we get an exception before this line. If `statistics_` is set, then that means we successfully ran `sync_with_server` and the publisher has stopped. If there _is_ a race, then I would expect there to be a race with the access to `collected_info_` in `sync_with_server`.

Reply:
> IIUC, the timer should already be canceled so we shouldn't be accessing this in another thread by this point

Yes, the mutex seems to be unnecessary in `ClientNode`.
There's a similar mutex in `ServerNode` though, and in that case it's necessary because `extract_results()` is called concurrently while an executor spins.
I could remove it from one and keep it in the other.

Reply:
What was the result of this comment chain?